### PR TITLE
Peer stats 2020 05 20

### PR DESCRIPTION
**Pull request by notlesh**
This PR adds a framework for collecting peer stats for use in analyzing performance of Service Node peers.

### CODE REVIEW

File: llarp/peerstats/peer_db.hpp [hunk 1]
```diff
@@ -62,11 +65,27 @@ namespace llarp
...
+    bool
+    shouldFlush(llarp_time_t now);
+
    private:
     std::unordered_map<RouterID, PeerStats, RouterID::Hash> m_peerStats;
     std::mutex m_statsLock;
 
     std::unique_ptr<PeerDbStorage> m_storage;
+
+    llarp_time_t m_targetFlushInterval = 30s;
```
Comment (majestrate) [inline llarp/peerstats/peer_db.hpp:87]: const?

Comment (notlesh) [reply llarp/peerstats/peer_db.hpp]: I might make it configurable, but I'll also make it `constexpr` until such time (d88ffbd)

File: llarp/peerstats/peer_db.cpp [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+
+namespace llarp
+{
+  PeerDb::PeerDb()
+  {
+    m_lastFlush.store({});
+  }
+
+  void
+  PeerDb::loadDatabase(std::optional<std::filesystem::path> file)
```
Comment (majestrate) [inline llarp/peerstats/peer_db.cpp:15]: use `fs::path` because macos is broke af

File: llarp/peerstats/peer_db.hpp [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    /// This must be called prior to calling flushDatabase(), and will truncate any existing data.
+    ///
+    /// This is a blocking call, both in the sense that it blocks on disk/database I/O and that it
+    /// will sit on a mutex while the database is loaded.
+    ///
+    /// @param file is an optional file which doesn't have to exist but must be writable, if a value
+    ///        is provided. If no value is provided, the database will be memory-backed.
+    /// @throws if sqlite_orm/sqlite3 is unable to open or create a database at the given file
+    void
+    loadDatabase(std::optional<std::filesystem::path> file);
```
Comment (majestrate) [inline llarp/peerstats/peer_db.hpp:38]: `fs::path` here too

Comment (notlesh) [reply llarp/peerstats/peer_db.hpp]: fixed (d2a27c9b)

File: llarp/peerstats/peer_db.cpp (hunk 2 above)
Comment (notlesh) [reply llarp/peerstats/peer_db.cpp]: fixed (d2a27c9b)

**Comment by majestrate** [comment]
i think we already merged this from another branch

**Pull request closed without merging**